bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/cli/Option.java

```diff
@@ -0,0 +1,222 @@
+        int index = 0;
+        while (index < args.length) {
+            String arg = args[index++];
+            if (CliUtil.isOption(arg)) {
+                index = handleOption(args, index, arg);
+            } else if (arg.startsWith("-")) {
+                skipConfigOption(arg);
+            } else {
+                operandArgs.add(arg);
+            }
+        }
+        validateRecordKeys();
+        return record;
+    }
+
+    private void skipConfigOption(String arg) {
+        // Skip the config options
+        if (arg.length() == 1 || arg.charAt(1) != 'C') {
+            throw ErrorCreator.createError(
+                    StringUtils.fromString("undefined cli argument: '" + arg + "'"));
+
+        }
+    }
+
+    private int handleOption(String[] args, int index, String optionStr) {
+        // Handle the case when there's only -- not followed by a string
+        if (optionStr.length() == 2) {
+            return addOperands(args, index);
+        }
+        BString optionName = StringUtils.fromString(getOptionName(optionStr));
+        validateRecordParam(optionName);
+        paramsFound.add(optionName);
+        if (isNamedArg(optionStr)) {
+            processNamedArg(optionStr, optionName);
+        } else {
+            if (!handleBoolean(optionName)) {
+                if (index < args.length) {
+                    handleOptionArgument(args[index++], optionStr, optionName);
+                } else {
+                    // case where the option argument is missing
+                    validateOptionArg(optionStr, optionStr);
+                }
+            }
+        }
+        return index;
+    }
+
+    private void handleOptionArgument(String val, String optionStr, BString optionName) {
+        Type fieldType = recordType.getFields().get(optionName.getValue()).getFieldType();
+        validateOptionArg(optionStr, val);
+        if (fieldType.getTag() == TypeTags.ARRAY_TAG) {
+            handleArrayParameter(optionName, val, (ArrayType) fieldType);
+        } else {
+            record.put(optionName, CliUtil.getBValue(fieldType, val));
+        }
+    }
+
+    private int addOperands(String[] args, int index) {
+        while (index < args.length) {
+            operandArgs.add(args[index++]);
+        }
+        return index;
+    }
+
+    private boolean handleBoolean(BString paramName) {
+        Type fieldType = recordType.getFields().get(paramName.getValue()).getFieldType();
+        if (isABoolean(fieldType)) {
+            record.put(paramName, true);
+            return true;
+        } else if (fieldType.getTag() == TypeTags.ARRAY_TAG) {
+            BArray bArray = getBArray(paramName, (ArrayType) fieldType);
+            Type elementType = bArray.getElementType();
+            if (isABoolean(elementType)) {
+                bArray.append(true);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void validateRecordKeys() {
```

Comment:
If we use mapvalueimpl constructor create with initial values we might be able to get rid of this check, since inside that we do the validation.